src/data/list/min_max.lean

```diff
@@ -271,4 +271,51 @@ theorem minimum_eq_coe_iff {m : α} {l : list α} :
   minimum l = m ↔ m ∈ l ∧ (∀ a ∈ l, m ≤ a) :=
 @maximum_eq_coe_iff (order_dual α) _ _ _
 
+section fold
+
+lemma maximum_eq_coe_foldr_max_of_ne_nil {α : Type*} [canonically_linear_ordered_add_monoid α]
+  (l : list α) (h : l ≠ []) :
+  l.maximum = (l.foldr max ⊥ : α) :=
+begin
+  induction l with hd tl IH,
+  { contradiction },
+  { rw [list.maximum_cons, list.foldr, with_bot.coe_max],
+    by_cases h : tl = [],
+    { simp [h, -with_top.coe_zero] },
+    { simp [IH h] } }
+end
+
+lemma maximum_nat_eq_coe_foldr_max_of_ne_nil (l : list ℕ) (h : l ≠ []) :
+  l.maximum = (l.foldr max 0 : ℕ) :=
+maximum_eq_coe_foldr_max_of_ne_nil l h
+
+lemma max_le_of_forall_le {α : Type*} [canonically_linear_ordered_add_monoid α]
+  (l : list α) (n : α) (h : ∀ (x ∈ l), x ≤ n) :
+  l.foldr max ⊥ ≤ n :=
+begin
+  induction l with y l IH,
+  { simp },
+  { specialize IH (λ x hx, h x (list.mem_cons_of_mem _ hx)),
+    have hy : y ≤ n := h y (list.mem_cons_self _ _),
+    simpa [hy] using IH }
+end
+
+lemma max_nat_le_of_forall_le (l : list ℕ) (n : ℕ) (h : ∀ (x ∈ l), x ≤ n) :
+  l.foldr max 0 ≤ n :=
+max_le_of_forall_le l n h
+
+@[to_additive]
+lemma prod_le_of_forall_le {α : Type*} [ordered_comm_monoid α]
```

Comment:
This feels out of place in this file

Reply:
Where would you put these then? I agree that it is a bit out of place because of how they are phrased. But it's not possible to state it like
```
l.prod ≤ (l.maximum) ^ l.length := sorry
```
without all sorts of extra typeclass assumptions.

Reply:
Bump

Reply:
I'd be tempted to just put this lemma in `list/basic`, since it's not about `min` or `max`.

Reply:
Obviously `basic` is huge, but that's an indication that it needs splitting into multiple files, not that we shouldn't put any new lemmas there.